### PR TITLE
Avoid O(n^2) lookup to remove duplicate subfolders in list output.

### PR DESCRIPTION
**Pull request by ncabatoff**

### CODE REVIEW

File: physical/raft/fsm.go [hunk 1]
```diff
@@ -352,7 +352,9 @@ func (f *FSM) List(ctx context.Context, prefix string) ([]string, error) {
 				keys = append(keys, key)
 			} else {
 				// Add truncated 'folder' paths
-				keys = strutil.AppendIfMissing(keys, string(key[:i+1]))
+				if len(keys) == 0 || keys[len(keys)-1] != key[:i+1] {
```
Comment (raskchanky) [inline physical/raft/fsm.go:355]: Is this safe, given that it's only ever looking at the last element?

Comment (ncabatoff) [reply physical/raft/fsm.go:355]: A question I've asked myself many times, and in fact this PR is my way of asking that.  As far as I can tell, the answer is yes.  The `Seek` we do above is premised on the bucket being ordered.

Comment (raskchanky) [reply physical/raft/fsm.go:355]: I think the part that keeps tripping me up when I think about this is the first `if` appending `key` to `keys` and the `else` appending `key[:i+1]` to `keys`. I don't know enough about this part of the code to know what those values typically look like, but the fact that they're slightly different, and this check isn't exhaustive, made me pause.

File: physical/raft/fsm_test.go [hunk 2]
```diff
@@ -125,3 +128,30 @@ func TestFSM_Batching(t *testing.T) {
...
+func TestFSM_List(t *testing.T) {
+	fsm, dir := getFSM(t)
+	defer os.RemoveAll(dir)
+
+	ctx := context.Background()
+	count := 100
+	keys := rand.Perm(count)
+	var sorted []string
+	for _, k := range keys {
+		err := fsm.Put(ctx, &physical.Entry{Key: fmt.Sprintf("foo/%d/bar", k)})
```
Comment (briankassouf) [inline physical/raft/fsm_test.go:141]: Could you write two values to this storage prefix so below we can make sure only one is listed?

Comment (ncabatoff) [reply physical/raft/fsm_test.go:141]: Done.

**Pull request merged**